File: Design Patterns/Command Pattern/command.ts
// This code shows the Command pattern in TypeScript.
// It defines a Light class (receiver), concrete command classes (TurnOnCommand, TurnOffCommand),
// and a RemoteControl class (invoker) that uses commands to control the light.

class Light {
  turnOn() {
    console.log("Light is on");
  }

  turnOff() {
    console.log("Light is off");
  }
}

interface Command {
  execute(): void;
}

class TurnOnCommand implements Command {
  constructor(private light: Light) {}

  execute() {
    this.light.turnOn();
  }
}

class TurnOffCommand implements Command {
  constructor(private light: Light) {}

  execute() {
    this.light.turnOff();
  }
}

class RemoteControl {
  private command: Command | null = null;

  setCommand(command: Command) {
    this.command = command;
  }

  pressButton() {
    if (this.command) {
      this.command.execute();
    } else {
      console.log("No command set");
    }
  }
}

// Usage
const light = new Light();
const turnOn = new TurnOnCommand(light);
const turnOff = new TurnOffCommand(light);
const remote = new RemoteControl();

remote.setCommand(turnOn);
remote.pressButton(); // Output: Light is on

remote.setCommand(turnOff);
remote.pressButton(); // Output: Light is off

File: Design Patterns/Command Pattern/command.py
# This code demonstrates the Command pattern in Python.
# It defines a Light class (receiver), concrete command classes (TurnOnCommand, TurnOffCommand),
# and a RemoteControl class (invoker) that uses commands to control the light.

from abc import ABC, abstractmethod

class Light:
    def turn_on(self):
        print("Light is on")

    def turn_off(self):
        print("Light is off")

class Command(ABC):
    @abstractmethod
    def execute(self):
        pass

class TurnOnCommand(Command):
    def __init__(self, light):
        self.light = light

    def execute(self):
        self.light.turn_on()

class TurnOffCommand(Command):
    def __init__(self, light):
        self.light = light

    def execute(self):
        self.light.turn_off()

class RemoteControl:
    def __init__(self):
        self.command = None

    def set_command(self, command):
        self.command = command

    def press_button(self):
        self.command.execute()

# Usage
light = Light()
turn_on = TurnOnCommand(light)
turn_off = TurnOffCommand(light)
remote = RemoteControl()

remote.set_command(turn_on)
remote.press_button()  # Output: Light is on

remote.set_command(turn_off)
remote.press_button()  # Output: Light is off